lib/exchanges/gdax.js

```diff
@@ -4,58 +4,78 @@ var request = require('request')
 var mapParse = require('./map-parsing.js')
 
 // Local variables
-var highbids
-var lowasks
+var orderbook;
+var min = .00000000001;
 
 // Function to parse the GDAX snapshot
-function parseSnap (data) {
-  var count = 100 // Number of orders to take from snapshot
-  for (var i = 0; i < count; i++) {
-    highbids.set(data.bids[i][2], {
-      rate: parseFloat(data.bids[i][0]),
-      amount: parseFloat(data.bids[i][1])
-    })
-    lowasks.set(data.asks[i][2], {
-      rate: parseFloat(data.asks[i][0]),
-      amount: parseFloat(data.asks[i][1])
-    })
-  }
+function parseSnap(data){
+    var count = 25; // Number of orders to take from snapshot
+    for (var i = 0; i < count; i++){
+      orderbook[parseFloat(data.bids[i][0])] = parseFloat(data.bids[i][1]);
+      orderbook[parseFloat(data.asks[i][0])] = -1 * parseFloat(data.asks[i][1]);
+    }
+
 }
 
 // Parse data from Web Socket
 // TODO: Explain (or link to) expected data scheme
-function parse (data) {
-  data = JSON.parse(data)
 
-  // Determine relevant order map to use
-  var bidsOrAsks = data.side === 'buy' ? highbids : lowasks
+function parse(data){
 
+  var data = JSON.parse(data)
+  var bidAskMult = 1;
+  // Simplifying for one orderbook template
+  if (data.side == "sell"){
+    bidAskMult = -1;
+  }
+  var key = parseFloat(data.price);
   switch (data.type) {
-    case 'open':
-      bidsOrAsks.set(data.order_id, {
-        rate: parseFloat(data.price),
-        amount: parseFloat(data.remaining_size)
-      })
-      break
-    case 'done':
-      bidsOrAsks.delete(data.order_id)
-      break
-    case 'change':
-      if (bidsOrAsks.get(data.order_id) !== undefined) {
-        bidsOrAsks(data.order_id).rate = parseFloat(data.price)
-        bidsOrAsks(data.order_id).amount = parseFloat(data.new_size)
-        bidsOrAsks(data.order_id).amount = parseFloat(data.new_funds)
+    case "open":
+      if(orderbook.hasOwnProperty(key)){
+        orderbook[key] += bidAskMult * parseFloat(data.remaining_size);
       }
-      break
+      else {
+        orderbook[key] = bidAskMult * parseFloat(data.remaining_size);
+      }
+      break;
+    case "done":
+      if (orderbook.hasOwnProperty(key)){
+        var tmp = parseFloat(orderbook[key]);
+        delete(orderbook[key]);
+        orderbook[key] = tmp - (bidAskMult * parseFloat(data.remaining_size));
+        if ((bidAskMult * orderbook[key]) < min){
+          delete(orderbook[key]);
+        }
+      }
+      break;
+    case "match":
+      if (orderbook.hasOwnProperty(key)){
+        var tmp = parseFloat(orderbook[key]);
+
+        delete(orderbook[key]);
+        orderbook[key] = tmp - (bidAskMult * parseFloat(data["size"]));
+
+        if ((bidAskMult * orderbook[key]) < min){
+          delete(orderbook[key]);
+        }
+      }
+      break;
+    case "change":
+      if (orderbook.hasOwnProperty(key)){
```

Comment:
The open case, reproduced here
```javascript
case "open":
      if(orderbook.hasOwnProperty(key)){
        orderbook[key] += bidAskMult * parseFloat(data.remaining_size);
      }
      else {
        orderbook[key] = bidAskMult * parseFloat(data.remaining_size);
      }
      break;
```
Means that you couldn't do such a blanket statement at the top of the switch statement, although we could maybe make two switch statements or I could just be wrong altogether.